### PR TITLE
tpl: Add uniq function

### PR DESCRIPTION
**Pull request by adiabatic**
As I [mentioned in the forums](https://discuss.gohugo.io/t/adding-uniq-to-template-functions/6710).

### CODE REVIEW

**Comment by adiabatic** [comment]
I made it fail nicely for `{1, 1, true},` and added support for arrays, although I'm not sure how they'd get passed in.

`{"foo", "fo", false},`, on the other hand, I changed to `{"foo", "fo", true},` because, on further reflection, I don't want this to operate on strings because it'll encourage normalization-dependent behavior. Depending on whether diacritical marks are part of the letter's code point or not, "foaö" could Uniq down to either "foaö" (if ö is one code point) or "foä" (if o and ¨ are separate code points).

File: tpl/collections/collections.go [hunk 1]
```diff
@@ -587,3 +587,48 @@ func (ns *Namespace) Union(l1, l2 interface{}) (interface{}, error) {
...
+	}
+
+	switch lv.Kind() {
+	case reflect.Array, reflect.Slice:
+	// okay
+	default:
+		return nil, errors.New("Can't uniquify a " + reflect.ValueOf(lv).Type().String())
+	}
+
+	var ret reflect.Value
```
Comment (moorereason) [inline tpl/collections/collections.go]: Can you refactor this a bit?  Define `ret` before the `switch` and merge the following `if` statements as cases in the `switch`.

File: tpl/collections/collections.go [hunk 2]
```diff
@@ -587,3 +587,48 @@ func (ns *Namespace) Union(l1, l2 interface{}) (interface{}, error) {
...
+// duplicate elements removed.
+func (ns *Namespace) Uniq(l interface{}) (interface{}, error) {
+	if l == nil {
+		return make([]interface{}, 0), nil
+	}
+
+	lv := reflect.ValueOf(l)
+	lv, isNil := indirect(lv)
+	if isNil {
+		return nil, errors.New("Can't uniquify a nil value")
```
Comment (moorereason) [inline tpl/collections/collections.go]: "uniquify" doesn't exist.  Change this to simply `invalid nil argument to Uniq`.

File: tpl/collections/collections.go [hunk 3]
```diff
@@ -587,3 +587,48 @@ func (ns *Namespace) Union(l1, l2 interface{}) (interface{}, error) {
...
+	}
+
+	switch lv.Kind() {
+	case reflect.Array, reflect.Slice:
+	// okay
+	default:
+		return nil, errors.New("Can't uniquify a " + reflect.ValueOf(lv).Type().String())
```
Comment (moorereason) [inline tpl/collections/collections.go]: Change to `can't use Uniq on`.

File: tpl/collections/collections.go [hunk 4]
```diff
@@ -587,3 +587,48 @@ func (ns *Namespace) Union(l1, l2 interface{}) (interface{}, error) {
...
+
+	for i := 0; i != lv.Len(); i++ {
+		lvv := lv.Index(i)
+		lvv, isNil := indirect(lvv)
+		if isNil {
+			continue
+		}
+
+		inRet := ns.In(ret.Interface(), lvv.Interface())
+		if !inRet {
```
Comment (moorereason) [inline tpl/collections/collections.go]: Simplify to

```go
if !ns.In(ret.Interface(), lvv.Interface()) {
```

File: docs/content/templates/functions.md [hunk 5]
```diff
@@ -373,6 +373,15 @@ e.g.
        {{ .Content }}
     {{ end }}
 
+
+### uniq
+
+Loops through an array and creates a new array with only the first occurrence of any given value.
```
Comment (moorereason) [inline docs/content/templates/functions.md]: Generally, try to make the docs description match the func's godoc comment.  We eventually plan to auto-generate these from godoc.

**Comment by adiabatic** [comment]
`make test` and `make test-race` work fine on my machine, but one of the Travis builds broke in `hugolib.TestShortcodeFigure` on linux. Is this something I should try to correct, or was this caused by someone else? I'm still new at using git with other people.

**Comment by moorereason** [comment]
This looks great, @adiabatic.  Thanks for persevering!

The failed Travis builds are unrelated to this PR.

**Pull request merged**